proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1160 @@
+* `trs.params.generatorKey` must have length 32.
+
+
+##### Execution
+
+Executing an update generator key transaction `trs` is done by calling `validators.updateGeneratorKey(address, trs.params.generatorKey)` where 
+`address` is the 20-byte address derived from trs.senderPublicKey.
+
+
+#### Vote
+
+Transactions executing this command have:
+
+* `moduleID = MODULE_ID_DPOS`
+* `commandID  = COMMAND_ID_VOTE`
+
+
+##### Parameters
+
+```java
+voteTransactionParams = {
+    "type": "object",
+    "required": ["votes"],
+    "properties": {
+        "votes": {
+            "type": "array",
+            "fieldNumber": 1,
+            "items": {
+                "type": "object",
+                "required": ["delegateAddress", "amount"],
+                "properties": {
+                    "delegateAddress" : {
+                        "dataType": "bytes", 
+                        "fieldNumber": 1
+                    },
+                    "amount": {
+                        "dataType": "sint64",
+                        "fieldNumber": 2
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+The verification and execution of this transaction is specified in [LIP 0023][LIP-0023-voteTrs]. 
+This specification is followed to implement the vote command with the following additional point:
+
+* when executing a self-vote (a vote with the `delegateAddress` equal to the address derived from the transaction public key), 
+  modify the `delegateStore(delegateAddress).selfVotes` property and the `voterStore(delegateAddress).sentVotes` entry corresponding to `delegateAddress` identically.
+
+
+#### Unlock
+
+Transactions executing this command have:
+
+* `moduleID = MODULE_ID_DPOS`
+* `commandID = COMMAND_ID_UNLOCK`
+
+
+##### Parameters
+
+The `params` property of unlock transactions is empty.
+
+
+#### Verification
+
+Unlock transactions do not trigger specific verifications.
+
+
+#### Execution
+
+When executing an unlock transaction `trs`, the following is done:
+
+```python
+senderAddress = address corresponding to trs.senderPublicKey
+height = height of the block including trs
+
+for each unlockObject in voterStore(senderAddress).pendingUnlocks:
+    if (hasWaited(unlockObject, senderAddress, height) 
+        and not isPunished(unlockObject, senderAddress, height)
+        and isCertificateGenerated(unlockObject)):
+        delete unlockObject from voterStore(senderAddress).pendingUnlocks
+        token.unlock(senderAddress, MODULE_ID_DPOS, TOKEN_ID_DPOS, unlockObject.amount)
+```
+
+The definition and rationale for the `isCertificateGenerated` function is part of [LIP "Introduce unlocking condition for incentivizing certificate generation"][LIP-incentivizeCertificateGeneration]. 
+The `hasWaited` and `isPunished` functions are defined below and are rationalized in [LIP 0023][LIP-0023-unlockRationale] and [LIP 0024][LIP-0024-rationale] respectively.
```

Comment:
Maybe it would be nice to still include the description of the input parameters, as it looks to me that the interface changed (I guess because it is easier with the new architecture?)

Reply:
done